# Lab 02 - Bài tập

---

## THÔNG TIN

- Họ tên: Nguyễn Văn Hùng
- MSV: 23668291
- Ngày hoàn thành: 11/1/2025

---
## NUMPY

Bạn hãy phát sinh ngẫu nhiên mảng ndarray có tên `arr` gồm 100 phần tử số nguyên có giá trị từ 0..1000. Thực hiện các thao tác sau:

### 1. Hiển thị các phần tử thỏa yêu cầu dưới đây

- 10 phần tử đầu tiên
- 5 phần tử cuối cùng
- Các phần tử thỏa mãn điều kiện chia hết cho 5 và lớn hơn 100

In [ ]:
import numpy as np
arr = np.random.randint(0, 1000, 100)
print(arr)
print("10 phần tử đầu tiên: ", arr[:10])
print("5 phần tử cuối cùng: ", arr[-5:])
print("Các phần tử thỏa mãn điều kiện chia hết cho 5 và lớn hơn 100: ", arr[(arr % 5 == 0) & (arr > 100)])

[567  98 597 581 339 910 189 637 699 758 145  67 159 634 804 357 559  72
 637 133 940 792 744 571 476 968 770 705 954 197 615 610 986 896 154 187
 411 306 232 553 618 376 112 667 397 602 466 144 547 826  80 725  89 712
 963 462 659 467 267 857 812 949  48 258 301 290 897 200 519 735 321 773
 181 589 955 306 426  59 782 853 389   7 796 878  85  73 606 201 677 885
 296 874 172 778 667 488 684 231 577 937]
10 phần tử đầu tiên:  [567  98 597 581 339 910 189 637 699 758]
5 phần tử cuối cùng:  [488 684 231 577 937]
Các phần tử thỏa mãn điều kiện chia hết cho 5 và lớn hơn 100:  [910 145 940 770 705 615 610 725 290 200 735 955 885]


### 2. Cho biết mảng trên có bao nhiêu phần tử có giá trị nhỏ hơn 100. Hiển thị các phần tử đó

In [ ]:
less_then_100 = arr[arr < 100]
print("Các phần tử có giá trị nhỏ hơn 100: ", less_then_100)
print(f"Có {len(less_then_100)} phần tử có giá trị nhỏ hơn 100!")

Các phần tử có giá trị nhỏ hơn 100:  [98 67 72 80 89 48 59  7 85 73]
Có 10 phần tử có giá trị nhỏ hơn 100!


### 3. Thực hiện thống kê:

- Tính tổng mảng
- Tìm giá trị trung bình
- Tìm giá trị trung vị
- Tìm Q1, Q2, Q3
- Tìm phân vị 20 (percentile 20)

In [ ]:
sum = np.sum(arr)
mean = np.mean(arr)
median = np.median(arr)
q1 = np.percentile(arr, 25)
q2 = np.percentile(arr, 50)
q3 = np.percentile(arr, 75)
percentile_20 = np.percentile(arr, 20)
print(f"Tổng mảng: {sum}")
print(f"Giá trị trung bình: {mean}")
print(f"Giá trị trung vị: {median}")
print(f"Q1: {q1}")
print(f"Q2: {q2}")
print(f"Q3: {q3}")
print(f"Phân vị 20: {percentile_20}")

Tổng mảng: 51600
Giá trị trung bình: 516.0
Giá trị trung vị: 569.0
Q1: 251.5
Q2: 569.0
Q3: 761.0
Phân vị 20: 195.4


---

## PANDAS

Thông tin các bộ dữ liệu:

- `car-sales.csv`: bộ dữ liệu chứa thông tin của các ôtô cũ ở một cửa hàng


### 1. Đọc dữ liệu

In [ ]:
# YOUR CODE HERE
import pandas as pd
import io
from google.colab import files
uploaded = files.upload()
df = pd.read_csv(io.BytesIO(uploaded['car-sales.csv']))
print(df)

Saving car-sales.csv to car-sales.csv
  Manufacturer Colour  Odometer  Doors    Price
0       Toyota  White    150043      4   4000.0
1        Honda    Red     87899      4   5000.0
2       Toyota   Blue     32549      3   7000.0
3          BMW  Black     11179      5  22000.0
4       Nissan  White    213095      4   3500.0
5       Toyota  Green     99213      4   4500.0
6        Honda   Blue     45698      4   7500.0
7        Honda   Blue     54738      4   7000.0
8       Toyota  White     60000      4   6250.0
9       Nissan  White     31600      4   9700.0


### 2. Dữ liệu có bao nhiêu dòng, bao nhiêu cột

Đáp án: (10 dòng, 5 cột)

In [ ]:
# YOUR CODE HERE
a=df.shape
print(f"Dữ liệu có {a[0]} dòng, {a[1]} cột")

Dữ liệu có 10 dòng, 5 cột


- Định tính (phân loại): Manufacturer, Colour
- Định lượng: Odemeter, Price, Doors

In [ ]:
# YOUR CODE HERE
qualitative_cols = ['Manufacturer', 'Colour']
quantitative_cols = ['Odometer', 'Price', 'Doors']

print("Qualitative columns:")
print(df[qualitative_cols].head(10))

print("\nQuantitative columns:")
print(df[quantitative_cols].head())

Qualitative columns:
  Manufacturer Colour
0       Toyota  White
1        Honda    Red
2       Toyota   Blue
3          BMW  Black
4       Nissan  White
5       Toyota  Green
6        Honda   Blue
7        Honda   Blue
8       Toyota  White
9       Nissan  White

Quantitative columns:
   Odometer    Price  Doors
0    150043   4000.0      4
1     87899   5000.0      4
2     32549   7000.0      3
3     11179  22000.0      5
4    213095   3500.0      4


### 3. Trong dữ liệu có thông tin của bao nhiêu xe?

Đáp án: 10 xe

In [ ]:
# YOUR CODE HERE
num_cars = df.shape[0]
print(f"Trong dữ liệu có thông tin của {num_cars} xe")

Trong dữ liệu có thông tin của 10 xe


### 4. Đếm số xe theo theo hãng

Đáp án: Toyota: 4 xe, Honda: 3 xe, Nissan: 2 xe, BMW: 1 xe

In [ ]:
# YOUR CODE HERE
num_cars_Manufacturer = df.groupby('Manufacturer').size()#df['Manufacturer'].value_counts()
print(num_cars_Manufacturer)

Manufacturer
BMW       1
Honda     3
Nissan    2
Toyota    4
dtype: int64


### 5. Đếm số xe theo màu


In [ ]:
num_cars_Colour = df['Colour'].value_counts()#df.groupby('Colour').size()
print(num_cars_Colour)

Colour
White    4
Blue     3
Red      1
Black    1
Green    1
Name: count, dtype: int64


### 6. Tổng giá trị các xe trong cửa hàng

Đáp án: 76450 USD


In [ ]:
# YOUR CODE HERE
sum_Price = df['Price'].sum()
print(f"{sum_Price} USD")

76450.0 USD


### 7. Top 3 xe có giá cao nhất


In [ ]:
# YOUR CODE HERE
print(df.sort_values(by = 'Price', ascending = False).head(3))


  Manufacturer Colour  Odometer  Doors    Price
3          BMW  Black     11179      5  22000.0
9       Nissan  White     31600      4   9700.0
6        Honda   Blue     45698      4   7500.0


### 8. Top 3 xe đã chạy nhiều cây số (km)nhất (Odometer)

Đáp án:
- Xe Nissan, White, 213095 km
- Xe Toyota, White, 150043 km
- Xe Toyota, Green, 99213 km

In [ ]:
# YOUR CODE HERE
print(df.sort_values(by = 'Odometer', ascending = False).head(3))

  Manufacturer Colour  Odometer  Doors   Price
4       Nissan  White    213095      4  3500.0
0       Toyota  White    150043      4  4000.0
5       Toyota  Green     99213      4  4500.0


### 9. Khảo sát xe của hãng Honda

- Tính tỷ lệ xe Honda trong dữ liệu của cửa hàng (0.3)
- Cho biết giá trung bình của xe hãng Honda của cửa hàng (6500)
- Cho biết các xe của hãng Honda trong cửa hàng có những màu sắc nào (Red, Blue)

In [ ]:
# YOUR CODE HERE
honda = df[df['Manufacturer'] == 'Honda']
print(honda)
print(f"Tỷ lệ xe Honda trong dữ liệu của cửa hàng: {len(honda) / len(df):.1f}")
print(f"Giá trung bình của xe hãng Honda của cửa hàng: {honda['Price'].mean():.2f}")
print(f"Các xe của hãng Honda trong cửa hàng có những màu sắc: {honda['Colour'].unique()}")

  Manufacturer Colour  Odometer  Doors   Price
1        Honda    Red     87899      4  5000.0
6        Honda   Blue     45698      4  7500.0
7        Honda   Blue     54738      4  7000.0
Tỷ lệ xe Honda trong dữ liệu của cửa hàng: 0.3
Giá trung bình của xe hãng Honda của cửa hàng: 6500.00
Các xe của hãng Honda trong cửa hàng có những màu sắc: ['Red' 'Blue']


### 10. Tính toán các giá trị sau của  thuộc tính giá ('Price')

- Giá trung bình các xe (7645)
- Giá trị trung vị các xe (6625)
- Miền giá trị của xe (18500)
- Miền phân vị của xe (2750)
- Điểm phân vị 0.4 và 0.6 của giá ([5750, 7000])

In [ ]:
# YOUR CODE HERE
print(f"Giá trung bình các xe: {df['Price'].mean():.2f}")
print(f"Giá trị trung vị các xe: {df['Price'].median():.2f}")
print(f"Miền giá trị của xe: {df['Price'].max() - df['Price'].min():.2f}")
print(f"Miền phân vị của xe: {np.percentile(df['Price'], 75) - np.percentile(df['Price'], 25):.2f}")
print(f"Điểm phân vị 0.4 và 0.6 của giá: ([{np.percentile(df['Price'], 40):.1f},{np.percentile(df['Price'], 60):.1f}])")

Giá trung bình các xe: 7645.00
Giá trị trung vị các xe: 6625.00
Miền giá trị của xe: 18500.00
Miền phân vị của xe: 2750.00
Điểm phân vị 0.4 và 0.6 của giá: ([5750.0,7000.0])


---